File: diskmond.pl
#!/usr/bin/perl -w
#
#########################################################################
#
# File: diskmond
#
# Purpose: diskmond checks every 60 seconds to make sure that the disk
#          utilization for the partition that holds the psad "fwdata"
#          file is not beyond a threshold that the administrator
#          defines.
#
# Author: Michael Rash (mbr@cipherdyne.com)
#
# Credits:  (see the CREDITS file)
#
# Version: 1.1.1
#
# Copyright (C) 1999-2002 Michael Rash (mbr@cipherdyne.com)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
#    USA
#
#########################################################################
#
# $Id$
#

use lib '/usr/lib/psad';
use Psad;
use Sys::Hostname 'hostname';
use File::stat;
use File::Copy;
use POSIX 'setsid';
use Getopt::Long 'GetOptions';
use strict;

### establish the default path to the config file (can be
### over-ridden with the -c <file> command line option.
my $CONFIG_FILE = '/etc/psad/psad.conf';

### handle command line arguments
die " ** Specify the path to the psad.conf file with " . 
    "\"-c <file>\".\n\n" unless (GetOptions (
    'config=s' => \$CONFIG_FILE
));

### read in the configuration file
my ($Config_href, $Cmds_href) = &Psad::buildconf($CONFIG_FILE);

### make sure the configuration is complete
&check_config();

my %Config = %$Config_href;
my %Cmds   = %$Cmds_href;
#my $Archive_files_aref = $Config{'ARCHIVE_FILES'};

### Check to make sure the commands specified in the config section
### are in the right place, and attempt to correct automatically if not.
&Psad::check_commands(\%Cmds);

&Psad::unique_pid($Config{'DISKMOND_PID_FILE'});

### install WARN and DIE handlers
$SIG{'__WARN__'} = \&Psad::warn_handler;
$SIG{'__DIE__'}  = \&Psad::die_handler;

my $pid = fork;
exit if $pid;
die " ** $0: Couldn't fork: $!" unless defined($pid);
POSIX::setsid() or die " ** $0: Can't start a new session: $!\n";

### write the pid to the pid file
&Psad::writepid($Config{'DISKMOND_PID_FILE'});

### initialize partition usage
my $usage     = 0;
my $email_ctr = 0;
my $hostname  = hostname;

my $config_mtime = stat($CONFIG_FILE)->mtime;
#====================== main =======================
### main loop
for (;;) {
    ### See if we need to import any changed config variables
    &check_import_config(\$config_mtime, $CONFIG_FILE);

    $usage = &get_usage();
    ### Check to see if we need to start archiving
    if ($usage >= $Config{'MAX_DISK_PERCENTAGE'}) {
        &rm_data();
        open DW, "> $Config{'PSAD_DIR'}/disk_limit";
        print DW "psad: disk utilization above $Config{'MAX_DISK_PERCENTAGE'}%\n";
        print DW "      on disk partition that holds $Config{'PSAD_DIR'}\n";
        print DW "      diskmond removed all ip directories in " .
            "$Config{'PSAD_DIR'}.\n";
        close DW;
        &Psad::sendmail("psad: disk utilization above $Config{'MAX_DISK_PERCENTAGE'}%",
            "$Config{'PSAD_DIR'}/disk_limit", $Config{'EMAIL_ADDRESSES'},
            $Cmds{'mail'});
        $email_ctr++;
        if ($email_ctr > 50) {
            &Psad::sendmail("psad: diskmond giving up, stopping psad.",
                $Config{'EMAIL_ADDRESSES'}, $Cmds{'mail'});
        }
    } else {
        $email_ctr = 0;
    }
    ### check disk usage every $CHECK_INTERVAL seconds
    sleep $Config{'DISKMOND_CHECK_INTERVAL'};
}
exit 0;
#===================== end main ====================

sub get_usage() {
    my @df_data = `$Cmds{'df'} $Config{'PSAD_DIR'}`;
    my ($prcnt) = ($df_data[$#df_data] =~ /(\d+)%/);
    return $prcnt;
}

sub rm_data() {
    chdir $Config{'PSAD_DIR'} or die
        " ** Could not chdir $Config{'PSAD_DIR'}: $!";

    &rm_scanlog($Config{'PSAD_DIR'});
    &rm_scanlog($Config{'SCAN_DATA_ARCHIVE_DIR'});

    if (-e $Config{'FW_DATA_FILE'}) {
        open F, "> $Config{'FW_DATA_FILE'}";
        close F;
    }
    if (-e "$Config{'SCAN_DATA_ARCHIVE_DIR'}/fwdata_archive") {
        open F, "> $Config{'SCAN_DATA_ARCHIVE_DIR'}/fwdata_archive";
        close F;
    }
    return;
}

sub rm_scanlog() {
    my $dir = shift;
    chdir $dir or die " ** Could not chdir($dir): $!\n";
    opendir D, $dir or
        die " ** Could not open dir: $dir: $!";
    my @files = readdir D;
    closedir D;
    shift @files; shift @files;
    for my $file (@files) {
        if ($file =~ /(?:\d{1,3}\.){3}\d{1,3}/ && -d $file) {
            ### we found a directory like /var/log/psad/<ip>, which
            ### contains the scanlog file for this ip
            if (-e "${file}/scanlog") {
                open F, "> ${file}/scanlog";
                close F;
            }
        }
    }
    return;
}

sub check_import_config() {
    my ($mtime_ref, $file) = @_;
    my $mtime_tmp = stat($file)->mtime;
    if ($mtime_tmp != $$mtime_ref) {  ### the file was modified, so import

        ($Config_href, $Cmds_href) = &Psad::buildconf($file);

        ### make sure the configuration is complete
        &check_config();

        %Config = %$Config_href;
        %Cmds   = %$Cmds_href;
        $$mtime_ref = $mtime_tmp;
    }
    return;
}

sub check_config() {
    my @required_vars = qw(
        FW_DATA DISKMOND_CHECK_INTERVAL
        MAX_DISK_PERCENTAGE DISKMOND_PID_FILE
        PSAD_DIR SCAN_DATA_ARCHIVE_DIR
    );
    &Psad::defined_vars($CONFIG_FILE, \@required_vars, $Config_href);
    return;
}

File: diskmond.c
/*
*****************************************************************************
*
*  File: diskmond.c
*
*  Purpose: diskmond checks the disk partition on which psad scan data
*           lives and will remove the data if the disk utilization goes
*           above a (configurable) threshold (95% by default).
*
*  Author: Michael B. Rash (mbr@cipherdyne.com)
*
*  Credits:  (see the CREDITS file)
*
*  Version: 1.0.0-pre4
*
*  Copyright (C) 1999-2001 Michael B. Rash (mbr@cipherdyne.com)
*
*  License (GNU Public License):
*
*     This program is distributed in the hope that it will be useful,
*     but WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*     GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
*     USA
*
*****************************************************************************
*
*  $Id$
*/

/* INCLUDES *****************************************************************/
#include "psad.h"

/* GLOBALS ******************************************************************/
short int email_ctr = 0;
const char mail_redr[] = " < /dev/null > /dev/null 2>&1";
const char hostname[] = HOSTNAME;
char mail_addrs[MAX_GEN_LEN+1];
char mailCmd[MAX_GEN_LEN+1];

/* PROTOTYPES ***************************************************************/
static void parse_config(
    char *config_file,
    char *mailCmd,
    char *mail_addrs,
    char *diskmond_pid_file,
    unsigned short int *max_disk_percentage,
    unsigned int *diskmond_check_interval,
    unsigned int *diskmond_max_retries
);

/* MAIN *********************************************************************/
int main(int argc, char *argv[]) {
    char config_file[MAX_PATH_LEN+1];
    char diskmond_pid_file[MAX_PATH_LEN+1];
    unsigned short int max_disk_percentage = 95; /* default to 95% utilization */
    unsigned int diskmond_check_interval = 5;  /* default to 5 seconds */
    unsigned int diskmond_max_retries = 10;    /* default to 10 tries */
    time_t config_mtime;
    struct stat statbuf;

#ifdef DEBUG
    printf(" ... Entering DEBUG mode ...\n");
    sleep(1);
#endif

    /* handle command line arguments */
    if (argc == 1) {  /* nothing but the program name was
                         specified on the command line */
        strcpy(config_file, CONFIG_FILE);
    } else if (argc == 2) {  /* the path to the config file was
                                supplied on the command line */
        strcpy(config_file, argv[1]);
    } else {
        printf(" ... You may only specify the path to a single config file:  ");
        printf("Usage:  psadwatchd <configfile>\n");
        exit(EXIT_FAILURE);
    }

    if (stat(config_file, &statbuf)) {
        printf(" ... @@@ Could not get mtime for config file: %s\n",
            config_file);
        exit(EXIT_FAILURE);
    }

    /* initialize config_mtime */
    config_mtime = statbuf.st_mtime;

#ifdef DEBUG
    printf(" ... parsing config_file: %s\n", config_file);
#endif

    /* parse the config file */
    parse_config(
        config_file,
        mailCmd,
        mail_addrs,
        disdmond_pid_file,
        &max_disk_percentage,
        &diskmond_check_interval,
        &diskmond_max_retries
    );

    /* first make sure there isn't another psadwatchd already running */
    check_unique_pid(psadwatchd_pid_file, "psadwatchd");

#ifndef DEBUG
    /* become a daemon */
    daemonize_process(psadwatchd_pid_file);
#endif

    /* start doing the real work now that the daemon is running and
     * the config file has been processed */

#ifdef DEBUG
    printf("\n");
#endif

    /* MAIN LOOP: */
    for (;;) {
        sleep(psadwatchd_check_interval);

        /* check to see if we need to re-import the config file */
        if (check_import_config(&config_mtime, config_file)) {
#ifdef DEBUG
    printf(" ... re-parsing config file: %s\n", config_file);
#endif
            /* reparse the config file since it was updated */
            parse_config(
                config_file,
                mailCmd,
                mail_addrs,
                diskmond_pid_file,
                &max_disk_percentage,
                &diskmond_check_interval,
                &diskmond_max_retries
            );
        }
    }

    /* this statements don't get executed, but for completeness... */
    exit(EXIT_SUCCESS);
}
/******************** end main ********************/

static void parse_config(
    char *config_file,
    char *mailCmd,
    char *mail_addrs,
    char *diskmond_pid_file,
    unsigned short int *max_disk_percentage,
    unsigned int *diskmond_check_interval,
    unsigned int *diskmond_max_retries)
{
    FILE *config_ptr;         /* FILE pointer to the config file */
    int linectr = 0;
    char config_buf[MAX_LINE_BUF];
    char char_diskmond_check_interval[MAX_NUM_LEN+1];
    char char_diskmond_max_retries[MAX_NUM_LEN+1];
    char *index;

    if ((config_ptr = fopen(config_file, "r")) == NULL) {
        perror(" ... @@@ Could not open config file");
        exit(EXIT_FAILURE);
    }

    /* increment through each line of the config file */
    while ((fgets(config_buf, MAX_LINE_BUF, config_ptr)) != NULL) {
        linectr++;
        index = config_buf;  /* set the index pointer to the
                                beginning of the line */

        /* advance the index pointer through any
         * whitespace at the beginning of the line */
        while (*index == ' ' || *index == '\t') index++;

        /* skip comments and blank lines, etc. */
        if ((*index != '#') && (*index != '\n')
                && (*index != ';') && (index != NULL)) {

            find_char_var("mailCmd ", mailCmd, index);
            find_char_var("EMAIL_ADDRESSES ", mail_addrs, index);
            find_char_var("DISKMOND_PID_FILE ", diskmond_pid_file, index);
            find_char_var("MAX_DISK_PERCENTAGE ",
                char_max_disk_percentage, index);
            find_char_var("DISKMOND_CHECK_INTERVAL ",
                char_diskmond_check_interval, index);
            find_char_var("DISKMOND_MAX_RETRIES ",
                char_diskmond_max_retries, index);
            find_char_var("DISKMOND_PID_FILE ", diskmond_pid_file, index);
        }
    }
    *max_disk_percentage     = atoi(char_max_disk_percentage);
    *diskmond_check_interval = atoi(char_diskmond_check_interval);
    *diskmond_max_retries    = atoi(char_diskmond_max_retries);
    fclose(config_ptr);
    return;
}
